### PR TITLE
Feature/unit tests/hdmap utils

### PR DESCRIPTION
**Pull request by robomic**
# Description

## Abstract

This PR contains unit tests for `HdMapUtils`, which is a part of traffic_simulator package.

## Details
Three customized maps have been added to the `map` folder in order to test some functions that require a specific setup.
All tests are included in `test_hdmap_utils.cpp` file.
In addition, already existing tests have also been moved to `test_hdmap_utils.cpp` file (tests 1-11).

Added maps:
1. empty
2. four_track_highway
3. crossroads_with_stoplines

List of tests included in this PR:
1. Construct
2. matchToLane
3. AlongLaneletPose
4. AlongLaneletPose_afterLast
5. AlongLaneletPose_beforeFirst
6. CanonicalizeNegative
7. CanonicalizePositive
8. Canonicalize
9. CanonicalizeAllNegative
10. CanonicalizeAllPositive
11. CanonicalizeAll
12. Construct_invalid
13. filterLaneletIds_emptyIds
14. filterLaneletIds_invalidIds
15. filterLaneletIds_correct
16. filterLaneletIds_invalidSubtype
17. getNearbyLaneletIds
18. getNearbyLaneletIds_unsuccessful
19. getNearbyLaneletIds_crosswalkUnsuccessful
20. getCollisionPointInLaneCoordinate_intersects
21. getCollisionPointInLaneCoordinate_disjoint
22. getCollisionPointInLaneCoordinate_invalidLanelet
23. getCollisionPointInLaneCoordinate_invalidCrosswalkLanelet
24. matchToLane_includeCrosswalk
25. matchToLane_noMatch
26. toLaneletPose_correct
27. toLaneletPose_negativeOffset
28. toLaneletPose_reverse
29. toLaneletPose_notOnLanelet
30. toLaneletPose_empty
31. toLaneletPose_boundingBoxMatchPrevious
32. getSpeedLimit_correct
33. getSpeedLimit_crosswalk
34. getSpeedLimit_empty
35. getClosestLaneletId_near
36. getClosestLaneletId_away
37. getClosestLaneletId_crosswalkCloserButExcluded
38. getClosestLaneletId_onlyCrosswalkNearButExcluded
39. getClosestLaneletId_emptyMap
40. getPreviousLaneletIds
41. getNextLaneletIds
42. getPreviousLaneletIds_RoadShoulder
43. getNextLaneletIds_RoadShoulder
44. getNextLaneletIds_multipleNext
45. getPreviousLaneletIds_multiplePrevious
46. getNextLaneletIds_direction
47. getPreviousLaneletIds_direction
48. isInRoute_empty
49. isInRoute_notOnRoute
50. isInRoute_onRoute
51. isInLanelet_correct
52. isInLanelet_after
53. isInLanelet_before
54. toMapPoints_correctPoints
55. toMapPoints_negativeS
56. toMapPoints_sLargerThanLaneletLength
57. toMapPoints_empty
58. toMapPose_onlyOffset
59. toMapPose_additionalRotation
60. toMapPose_negativeS
61. toMapPose_sLargerThanLaneletLength
62. getLaneChangeableLaneletId_straight
63. getLaneChangeableLaneletId_leftNoChangable
64. getLaneChangeableLaneletId_leftChangable
65. getLaneChangeableLaneletId_rightNoChangable
66. getLaneChangeableLaneletId_rightChangable
67. getLaneChangeableLaneletId_shift2LeftPossible
68. getLaneChangeableLaneletId_shift2LeftNotPossible
69. getLaneChangeableLaneletId_shift2RightPossible
70. getLaneChangeableLaneletId_shift2RightNotPossible
71. getLaneChangeableLaneletId_shift0
72. getTrafficLightIds_correct
73. getTrafficLightIds_noTrafficLight
74. getTrafficLightBulbPosition_correct
75. getTrafficLightBulbPosition_invalidTrafficLight
76. getConflictingLaneIds_conflicting
77. getConflictingLaneIds_notConflicting
78. getConflictingLaneIds_empty
79. getConflictingCrosswalkIds_conflicting
80. getConflictingCrosswalkIds_notConflictingWithCrosswalk
81. getConflictingCrosswalkIds_notConflicting
82. getConflictingCrosswalkIds_empty
83. clipTrajectoryFromLaneletIds_correct
84. clipTrajectoryFromLaneletIds_startNotOnTrajectory
85. clipTrajectoryFromLaneletIds_emptyTrajectory
86. clipTrajectoryFromLaneletIds_smallForwardDistance
87. getFollowingLanelets_straightAfter
88. getFollowingLanelets_curveAfter
89. getFollowingLanelets_notEnoughLaneletsAfter
90. getFollowingLanelets_candidateTrajectory
91. getFollowingLanelets_candidateTrajectoryNotEnough
92. getFollowingLanelets_candidateTrajectoryEmpty
93. getFollowingLanelets_candidatesDoNotMatch
94. canChangeLane_canChange
95. canChangeLane_canNotChange
96. canChangeLane_invalidLaneletId
97. getLateralDistance_sameLane
98. getLateralDistance_parallelLanesCanNotChange
99. getLateralDistance_parallelLanesCanChange
100. getLateralDistance_notConnected
101. getRoute_correct
102. getRoute_correctCache
103. getRoute_impossibleRouting
104. getRoute_circular
105. getCenterPoints_correct
106. getCenterPoints_correctCache
107. getCenterPoints_correctVector
108. getCenterPoints_empty
109. isTrafficLight_trafficLight
110. isTrafficLight_notTrafficLight
111. isTrafficLight_invalidId
112. isTrafficLightRegulatoryElement_trafficLightRegulatoryElement
113. isTrafficLightRegulatoryElement_noTrafficLightRegulatoryElement
114. isTrafficLightRegulatoryElement_invalidId
115. getLaneletLength_simple
116. getLaneletLength_cache
117. getTrafficLightIdsOnPath_trafficLights
118. getTrafficLightIdsOnPath_noTrafficLights
119. getTrafficLightIdsOnPath_empty
120. getLongitudinalDistance_sameLanelet
121. getLongitudinalDistance_differentLanelet
122. getLongitudinalDistance_sameLaneletBehind
123. getLongitudinalDistance_differentLaneletNoRoute
124. getStopLineIdsOnPath_someStopLines
125. getStopLineIdsOnPath_noStopLines
126. getStopLineIdsOnPath_empty
127. getTrafficLightStopLineIds_stopLine
128. getTrafficLightStopLineIds_severalStopLines
129. getTrafficLightStopLineIds_invalidTrafficLightId
130. getTrafficLightStopLinesPoints_stopLine
131. getTrafficLightStopLinesPoints_severalStopLines
132. getTrafficLightStopLinesPoints_invalidTrafficLightId
133. getStopLinePolygon_stopLine
134. getStopLinePolygon_invalidLaneletId
135. getDistanceToTrafficLightStopLine_trafficLightOnSpline
136. getDistanceToTrafficLightStopLine_noTrafficLightOnSpline
137. getDistanceToTrafficLightStopLine_trafficLightOnWaypoints
138. getDistanceToTrafficLightStopLine_noTrafficLightOnWaypoints
139. getDistanceToTrafficLightStopLine_emptyVector_waypoints
140. getDistanceToTrafficLightStopLine_routeWithTrafficLightsOnSpline
141. getDistanceToTrafficLightStopLine_routeWithNoTrafficLightsOnSplineCongruent
142. getDistanceToTrafficLightStopLine_routeWithTrafficLightsNotOnSplineIncongruent
143. getDistanceToTrafficLightStopLine_emptyVector_splineRoute
144. getDistanceToTrafficLightStopLine_routeWithTrafficLightsOnWaypoints
145. getDistanceToTrafficLightStopLine_routeWithNoTrafficLightsOnWaypointsIncongruent
146. getDistanceToTrafficLightStopLine_routeWithTrafficLightsNotOnWaypointsCongruent
147. getDistanceToTrafficLightStopLine_emptyVector_waypointsRoute
148. getDistanceToStopLine_stopLineOnSpline
149. getDistanceToStopLine_noStopLineOnSplineIncongruent
150. getDistanceToStopLine_noStopLineOnSplineCongruent
151. getDistanceToStopLine_emptyVector_spline
152. getDistanceToStopLine_stopLineOnWaypoints
153. getDistanceToStopLine_noStopLineOnWaypointsIncongruent
154. getDistanceToStopLine_noStopLineOnWaypointsCongruent
155. getDistanceToStopLine_emptyVector_waypoints

## References
Jira ticket: [internal link](https://tier4.atlassian.net/browse/RJD-1048)

# Destructive Changes

There are no destructive changes.

### CODE REVIEW

**Comment by robomic** [comment]
In [here](https://github.com/tier4/scenario_simulator_v2/pull/1270/commits/6a49e6b3570c6ba1dbbcd2fb2ac62bf4a8710238) this commit I fixed the test so that `include_crosswalk = true`.

In test `getStopLineIdsOnPath_someStopLines` I decided to alter the description for `a stop line` as the tested map only contains a singular stop sign with a stop line.

File: simulation/traffic_simulator/test/src/hdmap_utils/test_hdmap_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,3152 @@
...
+  const auto pose = makePose(makePoint(3790.0, 73757.0), makeQuaternionFromYaw(yaw));
+
+  const auto lanelet_pose = hdmap_utils.toLaneletPose(pose, false);
+
+  const auto reference_lanelet_pose =
+    traffic_simulator_msgs::build<traffic_simulator_msgs::msg::LaneletPose>()
+      .lanelet_id(id_road)
+      .s(35.0)
+      .offset(0.0)
+      .rpy(geometry_msgs::msg::Vector3());
```
Comment (hakuturu583) [inline simulation/traffic_simulator/test/src/hdmap_utils/test_hdmap_utils.cpp:657]: Temporary variables used only once should not be defined and should be assigned directly to arguments.

```suggestion
  const auto lanelet_pose = makeHdMapUtilsInstance().toLaneletPose(makePose(makePoint(3790.0, 73757.0), makeQuaternionFromYaw(M_PI + M_PI_2 / 3.0)), false);

  const auto reference_lanelet_pose =
    traffic_simulator_msgs::build<traffic_simulator_msgs::msg::LaneletPose>()
      .lanelet_id(34600)
      .s(35.0)
      .offset(0.0)
      .rpy(geometry_msgs::msg::Vector3());
```

Comment (hakuturu583) [reply simulation/traffic_simulator/test/src/hdmap_utils/test_hdmap_utils.cpp:657]: Please correct the other sections in the same manner.

File: simulation/traffic_simulator/test/src/hdmap_utils/test_hdmap_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,3152 @@
...
+/**
+ * @note Test basic functionality; test conversion to lanelet pose correctness with a point
+ * positioned on a given lanelet with the given matching distance
+ * - the goal is to test a regular usecase of correct conversion.
+ */
+TEST(HdMapUtils, toLaneletPose_correct)
+{
+  auto hdmap_utils = makeHdMapUtilsInstance();
+  const lanelet::Id id_road = 34600;
+  const double yaw = M_PI + M_PI_2 / 3.0;
```
Comment (hakuturu583) [inline simulation/traffic_simulator/test/src/hdmap_utils/test_hdmap_utils.cpp:647]: Please comment on why this hard-coded variable needs to be this value.

Comment (hakuturu583) [reply simulation/traffic_simulator/test/src/hdmap_utils/test_hdmap_utils.cpp:647]: Please correct the other sections in the same manner.

File: simulation/traffic_simulator/test/src/hdmap_utils/test_hdmap_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,3152 @@
...
+  origin.longitude = 139.78066608243;
+  EXPECT_THROW(hdmap_utils::HdMapUtils hdmap_utils(path, origin), std::runtime_error);
+}
+
+/**
+ * @note Test basic functionality; test map conversion to binary message correctness with a sample map.
+ */
+TEST(HdMapUtils, toMapBin)
+{
+  auto hdmap_utils = makeHdMapUtilsInstance();
```
Comment (hakuturu583) [inline simulation/traffic_simulator/test/src/hdmap_utils/test_hdmap_utils.cpp:119]: Almost all tests create an instance of HdMapUtils, but this item can be reduced in description by using Gtest's [TestFixture](https://google.github.io/googletest/primer.html#same-data-multiple-tests). Please use it to reduce the amount of description.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

**Comment by robomic** [comment]
@hakuturu583
I believe I addressed all the requests.

**Review by hakuturu583 — APPROVED** [review]

**Pull request merged**